### PR TITLE
Wraith Ability Refactor/Fixes

### PR DESCRIPTION
**Pull request by Surrealaser**
## About The Pull Request

1. Wraiths can no longer use Phase Shift while in the space of a phase blocking shutter, even if it's open.
2. Materializing in a solid object with Phase Shift no longer deducts plasma; it counts as if you had Resynced instead.
3. While the fight is Groundside Wraiths cannot phase shift into dropship area tiles from non-dropship area tiles.
4. General minor improvements to Wraith code.

## Why It's Good For The Game

Fixes some lingering issues with phase shift.

Improves Wraith code quality.

## Changelog
:cl:
fix: Wraiths can no longer use Phase Shift while in the space of a phase blocking shutter, even if it's open.
fix: Materializing in a solid object with Phase Shift no longer deducts plasma; it counts as if you had Resynced instead.
balance: While the fight is Groundside Wraiths cannot phase shift into dropship area tiles from non-dropship area tiles.
refactor: General minor improvements to Wraith code.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 1]
```diff
@@ -201,6 +204,17 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
...
 	var/phase_shift_duration_timer_id
+	var/phase_shift_duration_alert_id
+
+/datum/action/xeno_action/phase_shift/can_use_action(silent = FALSE, override_flags)
+	. = ..()
+	var/turf/T = get_turf(owner)
+	for(var/obj/machinery/door/poddoor/timed_late/containment/shutter_check AS in GLOB.wraith_no_incorporeal_pass_shutters)
+		if(locate(shutter_check) in T)
+			if(!silent)
+				to_chat(owner, span_xenowarning("We can't Phase Shift while in the space of warp protected shutters!"))
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:215]: Why not ust replace this with 
locate(/obj/machinery/door/poddoor/timed_late/containment) in T

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 2]
```diff
@@ -706,3 +714,30 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
 	banish_check.banish_deactivate()
 	succeed_activate()
 	add_cooldown()
+
+///Return TRUE if we have a block, return FALSE otherwise
+/atom/proc/turf_block_check(atom/target, ignore_can_pass = FALSE, ignore_density = FALSE, ignore_closed_turf = FALSE, ignore_invulnerable = FALSE, ignore_objects = FALSE)
+	var/turf/T = get_turf(target)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:720]: This should not be an atom level proc, it should be on the ability

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm (hunk 2 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:720]: It is called by multiple different abilities; how do you want this done exactly?

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm (hunk 1 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:215]: actually need to reduce the path to poddoor

Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:215]: The problem with using a simple locate is that there are two separate paths that need to be respected

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm]: What? locates check for children as well

Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm]: /obj/machinery/door/poddoor/timed_late/containment AND
/obj/machinery/door/poddoor/shutters/mainship/selfdestruct

I need to concern the check with these paths specifically; the common root of poddoor is too broad unfortunately.

Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm]: then why are you typecasting the for loop if its going to be typepaths not objects

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 3]
```diff
@@ -201,6 +202,17 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
 	var/turf/starting_turf = null
 	var/phase_shift_active = FALSE
 	var/phase_shift_duration_timer_id
+	var/phase_shift_duration_alert_id
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:205]: Autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 4]
```diff
@@ -201,6 +202,17 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
...
+
+/datum/action/xeno_action/phase_shift/can_use_action(silent = FALSE, override_flags)
+	. = ..()
+	var/turf/T = get_turf(owner)
+	for(var/obj/machinery/door/poddoor/shutter_check AS in GLOB.wraith_no_incorporeal_pass_shutters)
+		if(locate(shutter_check) in T)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:211]: Shouldn be typecasted

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 5]
```diff
@@ -727,3 +733,33 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
 	banish_check.banish_deactivate()
 	succeed_activate()
 	add_cooldown()
+
+///Return TRUE if we have a block, return FALSE otherwise
+proc/turf_block_check(atom/subject, atom/target, ignore_can_pass = FALSE, ignore_density = FALSE, ignore_closed_turf = FALSE, ignore_invulnerable = FALSE, ignore_objects = FALSE, ignore_mobs = FALSE, ignore_space = FALSE)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:738]: missingleading /

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 6]
```diff
@@ -174,8 +169,13 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
 	owner.forceMove(beacon_turf) //Move to where the beacon was
 	teleport_debuff_aoe(owner) //Apply tele debuff
 
+	var/warp_shadow_dissipate
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:172]:
```suggestion
	var/warp_shadow_dissipate = FALSE
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm (hunk 4 above)
Comment (Surrealaser) [reply code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:211]: Thought I fixed this, fuck

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**